### PR TITLE
[gemma4] Compute RMSNorm in fp32 to fix deep-layer hidden-state divergence vs HF

### PR DESCRIPTION
**Pull request by yihao-liang**
## Summary

gemma-4 (e.g. `google/gemma-4-31B-it`) produces hidden states that diverge from HuggingFace **only in the deep decoder layers**, collapsing to cosine ≈ 0.51 by layer 58 (same weights, bf16). Root cause: the fused CUDA `RMSNorm` reduces the variance in the **input (bf16) dtype**. gemma-4's deep layers grow massive-activation outlier dims, so the bf16 sum-of-squares loses precision; the per-layer error compounds across the 60-layer stack and collapses the deep hidden states.

This PR routes gemma-4's `RMSNorm`s through the fp32 `forward_native` path via a new `force_fp32` flag.

## Evidence (text GSM8K, cosine vs HF eager)

| layer | default (fused CUDA RMSNorm) | this PR (`force_fp32`) |
|------:|:--:|:--:|
| 52 | 0.913 | **0.999** |
| 57 | **0.744** | **0.999** |
| 58 | **0.513** | **0.999** |

A/B on the same engine isolates it precisely: forcing **only** the generic `RMSNorm` (sandwich + final norms; the q/k/v `Gemma4RMSNorm` left fused) onto `forward_native` recovers the deep cosine to ≥0.998. Cross-checks: a third engine (**vLLM**) matches HF ≥0.998 → SGLang-specific; **Qwen3.5** (which uses `GemmaRMSNorm`, not the generic `RMSNorm`) is unaffected; the vision encoder agrees ≥0.95 for all engines.

**Regression window:** introduced when the fp32 `forward_native` fallback was dropped from `RMSNorm.forward_cuda` — works at `692979a8d`, broken by `v0.5.12.post1` (`5a15cde`).

## Change

- `layers/layernorm.py`: add `force_fp32: bool = False` to `RMSNorm`; when set, dispatch to `forward_native` (fp32 variance reduction).
- `models/gemma4_causal.py`: pass `force_fp32=True` for gemma-4's RMSNorms.

## Notes

The deeper fix is to accumulate the variance in fp32 inside the fused `rmsnorm` / `fused_add_rmsnorm` kernels; this PR is a targeted, low-risk model-level fix (correctness ≫ a few % RMSNorm throughput on gemma-4). Marked **draft** pending maintainer CI / a GPU run — the numbers above are from an A/B that forces the same code path at runtime on the same engine.

## Scope note

The regression is in the **module-routed** norms: gemma-4's `input_layernorm`, `post_attention_layernorm`, `pre_feedforward_layernorm` and the final `norm` are called as `RMSNorm` modules, so `force_fp32` routes them to `forward_native` (fp32) — these are what the A/B fixes. The dense `post_feedforward_layernorm` (and the MoE branch) are instead computed via the separate `gemma_rmsnorm_residual_scalar` / `gemma_dual_rmsnorm_residual_scalar` Triton kernels, which are **not** part of this regression (unchanged across `692979a8d`→`v0.5.12`) and were verified not to be the cause. `force_fp32` on those modules still applies to their non-fused fallback branches; making those Triton kernels accumulate variance in fp32 can be a follow-up if full parity is wanted.

### CODE REVIEW

**Comment by yihao-liang** [comment]
### Independent GPU reproduction (released wheels, fresh H200)

Reproduced the A/B end-to-end on a clean H200 box, installing only the **released PyPI wheels** (`sglang==0.5.12.post1` + `sglang-kernel==0.4.2.post2`, torch 2.11) — no source build, no patches. The deep-layer collapse and the fp32-RMSNorm recovery match the table above:

```
[A FUSED  (default)]    L1=0.9979  L46=0.9895  L52=0.9129  L57=0.7443  L58=0.5127
forced forward_native (fp32) on 611 norm modules
[B NATIVE (force fp32)] L1=1.0     L46=0.9995  L52=0.9975  L57=0.9978  L58=0.9986
VERDICT: FUSED-CUDA RMSNorm IS THE REGRESSION
```

| layer | default (fused CUDA RMSNorm) | force fp32 (`forward_native`) |
|------:|:--:|:--:|
| 52 | 0.913 | 0.998 |
| 57 | **0.744** | 0.998 |
| 58 | **0.513** | 0.999 |

`L1 ≈ 0.998` in both runs is the alignment sanity check (capture vs HF).

**Standalone repro:** https://github.com/yihao-liang/sglang-gemma4-rmsnorm-repro

```bash
uv venv --python 3.12 .venv && . .venv/bin/activate
uv pip install --prerelease=allow -r requirements.txt
CUDA_VISIBLE_DEVICES=0 python repro_gemma4_e2e.py --n 16
```

It uses only public `sglang` + `transformers` APIs: an in-process `ModelRunner`, HF reference via `output_hidden_states`, and per-layer capture through plain PyTorch forward hooks on `language_model.layers`. The A and B runs use the same engine/weights/prompts — the only change is `_forward_method = forward_native` on the `RMSNorm` modules, i.e. fp32 vs bf16 variance reduction.

Note: a single-op `RMSNorm` micro-benchmark does **not** show the regression — `forward_cuda` and `forward_native` are bit-identical per call, because the per-op error is a near-uniform scale error (cosine is scale-invariant) below the bf16 output-quantization floor. It only emerges by compounding the systematic bias across the 60-layer residual stream, which is why the end-to-end A/B is the right isolation.

**Comment by pyc96** [comment]
I took a deeper look, I think the issue is a red herring coming from how the hidden states are captured in repro script. 

The model loop returns each layer's hidden_states as the next layer's input. 
Layer L's hook captures out[0] via .detach() → shares storage with layer L's output buffer.
That same buffer becomes layer L+1's residual, which fused_add_rmsnorm overwrites in place.

To verify, you can change your script hook from .detach() to .detach().clone().

FYI @kpham-sgl

**Comment by yihao-liang** [comment]
@pyc96 You're right — confirmed. This is a capture artifact in the repro, not a real RMSNorm regression.

**Mechanism** (exactly as you described): each gemma-4 layer returns `(hidden_states, None)` and the model loop feeds that tensor straight into the next layer as its `residual`. The hook's `out[0].detach()` shares storage with it, and the next layer's `pre_feedforward_layernorm` → `RMSNorm.forward_cuda` → `fused_add_rmsnorm` overwrites that buffer **in place**. `forward_native` adds **out-of-place** and rebinds, so it never mutates the captured tensor. That asymmetry — not fp32-vs-bf16 variance — is the entire A/B gap: run A's deep captures were corrupted after the fact; run B's weren't.

**Empirical check** on `google/gemma-4-31B-it` (single H200, sglang `v0.5.12.post1` code paths + transformers 5.8.1, 60 layers, n=4), two hooks per layer — `out[0].detach()` (alias) and `out[0].detach().clone()` (snapshot):

| layer | `cos(alias, clone)` FUSED | `cos(alias, clone)` NATIVE | `cos(clone_fused, clone_native)` |
|---:|:--:|:--:|:--:|
| 46 | 0.9906 | 1.0000 | 0.9999 |
| 52 | 0.9297 | 1.0000 | 0.9997 |
| 58 | **0.5792** | 1.0000 | **0.9997** |

- Col 1 reproduces the "deep collapse" — but it's purely the bare-`.detach()` buffer being overwritten in place, and only in the fused run.
- Col 2 is 1.0 everywhere: `forward_native` never mutates the capture, which is why run B looked "recovered."
- Col 3: with correct (`clone`) capture, fused and native hidden states are identical to ≈0.9997 through L58 → no real divergence. Since the fp32/native path already matches HF, the fused path does too.

So `force_fp32` isn't fixing a real bug. I've fixed the repro to use `.detach().clone()` and I'll close this PR. Good catch.
